codec/src/main/java/io/netty/handler/codec/CodecOutList.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec;
+
+import io.netty.util.Recycler;
+
+import java.util.AbstractList;
+import java.util.RandomAccess;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Special {@link AbstractList} implementation which is used within our codec base classes.
+ */
+final class CodecOutList extends AbstractList<Object> implements RandomAccess {
+
+    private static final Recycler<CodecOutList> RECYCLER = new Recycler<CodecOutList>() {
+        @Override
+        protected CodecOutList newObject(Handle<CodecOutList> handle) {
+            return new CodecOutList(handle);
+        }
+    };
+
+    static CodecOutList newInstance() {
+        return RECYCLER.get();
+    }
+
+    private final Recycler.Handle<CodecOutList> handle;
+    private int size;
+    // Size of 16 should be good enough for 99 % of all users.
```

Comment:
what about `8`? or something else? Do we have a good feel for what our codecs generally require for decoding?